lib/property_constraints.js

```diff
@@ -1,95 +1,116 @@
+  return function(other) {
+    return other === value;
+  };
+}
 
-  withSome: function(propertyValue, expectedSet) {
-    expectedSet = handleFunctionParameter(expectedSet, propertyValue);
-    return Immutable.Set(propertyValue).intersect(expectedSet).size > 0;
-  },
-  withNone: function(propertyValue, expectedSet) {
-    expectedSet = handleFunctionParameter(expectedSet, propertyValue);
-    return Immutable.Set(propertyValue).intersect(expectedSet).size === 0;
-  },
+function gt(value) {
+  return function(other) {
+    return other > value;
+  };
+}
 
-  withAll: function(propertyValue, expectedSet) {
-    expectedSet = handleFunctionParameter(expectedSet, propertyValue);
-    return Immutable.Set(expectedSet).intersect(propertyValue).size === expectedSet.size;
-  },
+function lt(value) {
+  return function(other) {
+    return other < value;
+  };
+}
 
-  withValueFrom: function(propertyValue, expectedSet) {
-    expectedSet = handleFunctionParameter(expectedSet, propertyValue);
-    return propertyValue.size === expectedSet.size && this.withAll(propertyValue, expectedSet);
-  }
+function filterWith(pred) {
+  return function(list) {
+    return list.filter(pred).length > 0;
+  };
+}
 
+function withSome(value) {
+  return function(list) {
+    return Immutable.Set(list).intersect(value).size > 0;
+  };
+}
 
-};
 
-module.exports = function(constraints) {
-  constraints = Immutable.fromJS(constraints);
-  return function(model) {
-    model = Immutable.fromJS(model);
-    var applyConstraint = function(constraint, model) {
-      if (constraint.get('hasNoProperty')) {
-        return !!!rules.hasProperty(model, constraint.get('hasNoProperty'));
-      }
-      var propertyName = constraint.get('hasProperty');
-      if (propertyName) {
-        var propertyValue = rules.hasProperty(model, propertyName);
-        var keys = constraint.keySeq().toArray()
-        .filter(function(key) {
-          return key !== 'hasProperty';
-        });
-        return !!(keys.reduce(function(mem, curr) {
-          return mem && rules[curr].apply(rules, [propertyValue, constraint.get(curr)]);
-        }, propertyValue));
-      }
-      return false;
-    };
-
-
-    if (constraints.size > 0) {
-      return constraints.slice(1).reduce(function(mem, curr) {
-        return mem && applyConstraint(curr, model);
-      }, applyConstraint(constraints.first(), model));
-    } else {
-      return true;
-    }
+function withAll(value) {
+  return function(list) {
+    return Immutable.Set(list).intersect(value).size === value.length;
   };
+}
+
+function isEmpty() {
+  return function(list) {
+    return list.length === 0;
+  };
+}
+
+var rules = {
+  constraint: constraint,
+  eq: eq,
+  gt: gt,
+  lt: lt,
+  and: and,
+  not: not,
+  hasProperty: hasProperty,
+  get: get,
+  getIn: getIn,
+  filterWith: filterWith,
+  withSome: withSome,
+  withAll: withAll,
+  isEmpty: isEmpty
+};
+
+module.exports = function(fn) {
+  return fn.apply(rules);
```

Comment:
Was this to avoid having to import a bunch of really small functions?

It would probably be more idiomatic for consumers to import the namespace with a short identifier e.g.

```javascript
var cp = require('./lib/property_constraints');

var predicate = cp.and(cp.get('foo'), cp.lt('5'));
```

Then in property constraints you just have multiple exports e.g.

```javascript
module.exports.gt = function(value) {
  return function(other) {
    return other > value;
  };
}

module.exports.and = function() {
//etc
}
```

Also tidies up some of the 'this' business.